### PR TITLE
MAINT: Avoid unnecessary array copies in xdist

### PR DESCRIPTION
**Pull request by deasmhumhna**
Striding problems are already dealt with by the call to np.ascontinguousarray in _convert_to_type. This avoids making unnecessary copies if X, etc. are contiguous views.

### CODE REVIEW

**Comment by perimosocordiae** [comment]
Failing tests look unrelated:
```
FAIL scipy/special/tests/test_basic.py::TestBessel::()::test_yv_cephes_vs_amos
FAIL scipy/special/tests/test_basic.py::TestBessel::()::test_iv_cephes_vs_amos
```

**Comment by perimosocordiae** [comment]
This change looks reasonable, though I'm not sure we have test cases for non c-contiguous input arrays.

**Comment by dezmondgoff** [comment]
I'll write a test and check it out. Travis build failed, but that doesn't seem to have anything to do with this.

File: scipy/spatial/tests/test_distance.py [hunk 1]
```diff
@@ -609,6 +609,31 @@ def test_cdist_out(self):
...
+        assert_(not X1.flags.c_contiguous)
+        assert_(not X2.flags.c_contiguous)
+        assert_(X1_copy.flags.c_contiguous)
+        assert_(X2_copy.flags.c_contiguous)
+
+        for metric in _METRICS_NAMES:
+            kwargs = dict()
+            if metric in ['minkowski', 'wminkowski']:
+                kwargs['p'] = 1.23
+            if metric == 'wminkowski':
```
Comment (perimosocordiae) [inline scipy/spatial/tests/test_distance.py]: This if statement can be indented; it'll only be true if the previous conditional was true.

File: scipy/spatial/tests/test_distance.py [hunk 2]
```diff
@@ -609,6 +609,31 @@ def test_cdist_out(self):
...
+                kwargs['w'] = 1.0 / X1.std(axis=0)
+            Y1 = cdist(X1, X2, metric, **kwargs)
+            Y2 = cdist(X1_copy, X2_copy, metric, **kwargs)
+            # test that output is numerically equivalent
+            _assert_within_tol(Y1, Y2, eps, verbose > 2)
```
Comment (perimosocordiae) [inline scipy/spatial/tests/test_distance.py]: Shouldn't these arrays be exactly equivalent?

**Review by perimosocordiae — APPROVED** [review]
Looks good overall, just a few minor comments.

File: scipy/spatial/tests/test_distance.py (hunk 2 above)
Comment (dezmondgoff) [reply scipy/spatial/tests/test_distance.py]: Yes. Are you suggesting to use _assert_equal() instead?

**Comment by perimosocordiae** [comment]
This PR looks to be essentially done. Anyone have further comments before I merge it?

**Review by tylerjereddy — COMMENTED** [review]
Looks like some low level changes covered by new unit tests with all tests passing. The one thing I'd suggest is `asv` benchmarks against the current master branch to show that performance improves (or at least remains constant) as a result of removing extraneous copy operations, as one might expect. Perhaps not crucial though if you're satisfied this is a clear improvement.

**Comment by apbard** [comment]
for what is worth... LGTM. :)

(remember to squash the commits)

**Comment by perimosocordiae** [comment]
Merged, sorry for the delay. Thanks, @dezmondgoff!

**Pull request merged**